shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -399,136 +433,447 @@ class Context {
-//   where only a single grain's shares need to be considered, it feels weird to think of the
-//   grain ID as being the primary distinguishing feature of the vertex.
+  activateOwnerEdges(grainId, edges) {
+    check(edges, [{ identityId: String, role: SandstormDb.prototype.roleAssignmentPattern }]);
 
-function computePermissionFlowToVertex(context, vertex, permissionSet, viewInfo) {
-  // Computes the flow of the permissions in `permissionSet` from the grain owner to `vertex`.
-  // Returns a PermissionFlow representing all permissions flowing into the vertex.
-  //
-  // That is to say, this computes the permissions held by a particular user or token (`vertex`)
-  // by traversing the sharing graph. The results may be conditional -- we compute permissions
-  // held under varying sets of conditions, and return a map of condition -> permissions. The
-  // caller may need to check if the conditions hold in order to compute the final permissions.
+    edges.forEach((edge) => {
+      const viewInfo = this.grains[grainId].cachedViewInfo || {};
+      const permissions = PermissionSet.fromRoleAssignment(edge.role, viewInfo);
+      const vertexId = "i:" + edge.identityId;
+      forEachPermission(permissions.array, (permissionId) => {
+        this.setToTrueStack.push({ grainId: grainId, vertexId: vertexId,
+                                   permissionId: permissionId, });
+      });
+    });
+  }
+
+  activateToken(tokenId) {
+    // Includes a new token (which must already be in `this.tokensById`) in our computation. The
+    // tricky part here is dealing with our already-accumulated knowledge; we need to compute how
+    // many of the token's requirements are currently unmet and whether we need to push anything new
+    // onto `setToTrueStack`.
+
+    check(tokenId, String);
+    if (tokenId in this.activeTokens) {
+      return false;
+    }
+
+    const token = this.tokensById[tokenId];
+    const grainId = token.grainId;
+    const viewInfo = this.grains[grainId].cachedViewInfo || {};
+    const tokenPermissions = PermissionSet.fromRoleAssignment(token.roleAssignment, viewInfo);
+
+    const sharerId = token.parentToken ? "t:" + token.parentToken : "i:" + token.identityId;
+    const recipientId = vertexIdOfTokenOwner(token);
+
+    let numUnmetRequirements = 0;
+    if (token.requirements) {
+      token.requirements.forEach((requirement) => {
+        if (requirement.permissionsHeld) {
+          const reqGrainId = requirement.permissionsHeld.grainId;
+          const reqVertexId = "i:" + requirement.permissionsHeld.identityId;
+          const reqPermissions = requirement.permissionsHeld.permissions || [];
+
+          forEachPermission(reqPermissions, (permissionId) => {
+            const variable = this.getVariable(reqGrainId, reqVertexId, permissionId);
+            if (!variable.value) {
+              numUnmetRequirements += 1;
+              variable.requirementsTailList.push(tokenId);
+            }
+          });
+        }
```

Comment:
A `tokenValid` requirement comes about when a capability was transmitted across a non-UiView capability. That is, the `tokenValid` requirement should always point to a non-UiView token. The requirement passes if restoring that token would currently succeed. This requires checking that the token and its ancestors (parentToken chain) all exist and are not revoked, and that all of their respective requirements are satisfied.